ReactWindows/ReactNative.Shared/UIManager/AccessibleAutomationPeer.cs

```diff
@@ -0,0 +1,104 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using ReactNative.UIManager.Events;
+#if WINDOWS_UWP
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Automation.Peers;
+using Windows.UI.Xaml.Automation.Provider;
+#else
+using System.Windows;
+using System.Windows.Automation.Peers;
+using System.Windows.Automation.Provider;
+#endif
+
+namespace ReactNative.UIManager
+{
+    /// <summary>
+    /// Automation peer for a FrameworkElement that implements IAccessible.
+    /// </summary>
+    /// <typeparam name="T">Type of owner of the automation peer.</typeparam>
+    public sealed class AccessibleAutomationPeer<T> : FrameworkElementAutomationPeer, IInvokeProvider
+        where T : FrameworkElement, IAccessible
+    {
+        /// <summary>
+        /// Hides base UIElement Owner to provide stronger-typed T Owner.
+        /// </summary>
+        private new T Owner => (T)base.Owner;
+
+        /// <inheritdoc />
+        public AccessibleAutomationPeer(T owner)
+            : base(owner)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override IList<AutomationPeer> GetChildrenCore()
+        {
+            bool shouldHideChildren =
+                Owner.ImportantForAccessibility == ImportantForAccessibility.Yes
+                || Owner.ImportantForAccessibility == ImportantForAccessibility.NoHideDescendants;
+            return shouldHideChildren ? null : base.GetChildrenCore();
+        }
+
+        /// <inheritdoc />
+        protected override bool IsContentElementCore()
+        {
+            return Owner.ImportantForAccessibility == ImportantForAccessibility.Yes;
+        }
+
+        /// <inheritdoc />
+        protected override bool IsControlElementCore()
+        {
+            return Owner.ImportantForAccessibility == ImportantForAccessibility.Yes;
+        }
+
+        /// <inheritdoc />
+        protected override AutomationControlType GetAutomationControlTypeCore()
+        {
+            if (Owner.AccessibilityTraits?.Contains(AccessibilityTrait.Button) == true)
+            {
+                return AutomationControlType.Button;
+            }
+            return base.GetAutomationControlTypeCore();
+        }
+
+        /// <inheritdoc />
+        protected override string GetNameCore()
+        {
+            string baseName = base.GetNameCore();
+            if (!string.IsNullOrWhiteSpace(baseName)
+                || Owner.ImportantForAccessibility != ImportantForAccessibility.Yes)
+            {
+                return baseName;
+            }
+
+            var sb = new StringBuilder();
+            foreach (var peer in base.GetChildrenCore())
```

Comment:
>base.GetChildrenCore [](start = 33, length = 20)

Should we call this class's GetChildren() method here instead?

Reply:
we cannot since then our override of GetChildrenCore will be called, and if importantForAccessibility==Yes or NoHideDescendants then it will return null. So we must call original base implementation via GetChildrenCore

---
In reply to: [159566940](https://github.com/reseul/react-native-windows/pull/1#discussion_r159566940) [](ancestors = 159566940)

Reply:
It looks like you decided to call `peer.GetChildren()`. Do you have to guard against that returning `null`?

---
In reply to: [159784879](https://github.com/reseul/react-native-windows/pull/1#discussion_r159784879) [](ancestors = 159784879,159566940)

Reply:
It looks like we need to call `peer.GetChildren()` in one case and `base.GetNameCore()` in another.

---
In reply to: [160286705](https://github.com/reseul/react-native-windows/pull/1#discussion_r160286705) [](ancestors = 160286705,159784879,159566940)

Reply:
great catch! made the recursive method parametrized not by peer but by list of peers (children).

---
In reply to: [160302736](https://github.com/reseul/react-native-windows/pull/1#discussion_r160302736) [](ancestors = 160302736,160286705,159784879,159566940)